backend/de.metas.ui.web.base/src/main/java/de/metas/ui/web/replenish/process/WEBUI_M_Replenish_Generate_Forecasts.java

```diff
@@ -0,0 +1,154 @@
+import de.metas.quantity.Quantitys;
+import de.metas.ui.web.process.adprocess.ViewBasedProcessTemplate;
+import de.metas.uom.UomId;
+import de.metas.util.Services;
+import lombok.NonNull;
+import org.adempiere.warehouse.api.IWarehouseBL;
+import org.compiere.SpringContextHolder;
+import org.compiere.model.I_M_Forecast;
+import org.compiere.util.Env;
+import org.compiere.util.TimeUtil;
+
+import javax.annotation.Nullable;
+import java.sql.Timestamp;
+import java.time.DayOfWeek;
+import java.time.LocalDate;
+import java.time.ZoneId;
+import java.time.temporal.TemporalAdjusters;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class WEBUI_M_Replenish_Generate_Forecasts extends ViewBasedProcessTemplate implements IProcessDefaultParametersProvider, IProcessPrecondition
+{
+	public static final AdMessageKey ALL_SELECTED_RECORDS_SHALL_HAVE_WAREHOUSE_AND_DEMAND_MSG = AdMessageKey.of("ALL_SELECTED_RECORDS_SHALL_HAVE_WAREHOUSE_AND_DEMAND_MSG");
+
+	@NonNull private final ForecastService forecastService = SpringContextHolder.instance.getBean(ForecastService.class);
+	@NonNull private final IProductBL productBL = Services.get(IProductBL.class);
+	@NonNull private final IOrgDAO orgDAO = Services.get(IOrgDAO.class);
+	@NonNull private final IWarehouseBL warehouseBL = Services.get(IWarehouseBL.class);
+
+	private static final String PARAM_DatePromised = I_M_Forecast.COLUMNNAME_DatePromised;
+	@Param(parameterName = PARAM_DatePromised, mandatory = true)
+	private Timestamp p_DatePromised;
+
+	@Override
+	public final ProcessPreconditionsResolution checkPreconditionsApplicable()
+	{
+		if (getSelectedRowIds().isEmpty())
+		{
+			return ProcessPreconditionsResolution.rejectBecauseNoSelection();
+		}
+
+		final boolean allRowsHaveDemand = streamSelectedRows()
+				.map(MaterialNeedsPlannerRow::ofRow)
+				.allMatch(MaterialNeedsPlannerRow::isDemandFilled);
+
+		if (!allRowsHaveDemand)
+		{
+			return ProcessPreconditionsResolution.reject(msgBL.getTranslatableMsgText(ALL_SELECTED_RECORDS_SHALL_HAVE_WAREHOUSE_AND_DEMAND_MSG));
+		}
+
+		return ProcessPreconditionsResolution.accept();
+	}
+
+	@Override
+	protected String doIt() throws Exception
+	{
+		final List<ForecastRequest> forecastRequests = streamSelectedRows()
+				.map(MaterialNeedsPlannerRow::ofRow)
+				.filter(MaterialNeedsPlannerRow::isDemandFilled)
+				.collect(Collectors.groupingBy(MaterialNeedsPlannerRow::getWarehouseId))
+				.entrySet()
+				.stream()
+				.map(entry -> ForecastRequest.builder()
+						.name(warehouseBL.getWarehouseName(entry.getKey()) + "_" + p_DatePromised)
+						.warehouseId(entry.getKey())
+						.datePromised(p_DatePromised)
+						.forecastLineRequests(entry.getValue().stream()
+								.map(this::toForecastLineRequest)
+								.collect(ImmutableList.toImmutableList()))
+						.build())
+				.collect(ImmutableList.toImmutableList());
+
+		final ImmutableSet<ForecastId> createdForecastIds = forecastService.createForecasts(forecastRequests);
```

Comment:
why not creating and completing the forecast in one go?

Reply:
IMHO they are two different operations, creating and completing so I went with two methods for a bigger flexibility of the service